cruise-control-core/src/main/java/com/linkedin/cruisecontrol/monitor/sampling/aggregator/AggregatedMetricValues.java

```diff
@@ -0,0 +1,127 @@
+                                               + "different lengths of " + length + " and " + values.length());
+      }
+    }
+    _metricValues = metricValues;
+  }
+  
+  public MetricValues valuesFor(int metricId) {
+    return _metricValues.get(metricId);
+  }
+
+  public int length() {
+    return _metricValues.isEmpty() ? 0 : _metricValues.values().iterator().next().length();
+  }
+
+  public boolean isEmpty() {
+    return _metricValues.isEmpty();
+  }
+  
+  public Set<Integer> metricIds() {
+    return Collections.unmodifiableSet(_metricValues.keySet());
+  }
+  
+  public void add(int metricId, MetricValues metricValuesToAdd) {
+    if (metricValuesToAdd == null) {
+      throw new IllegalArgumentException("The metric values to be added cannot be null");
+    }
+    if (!_metricValues.isEmpty() && metricValuesToAdd.length() != length()) {
+      throw new IllegalArgumentException("The existing metric length is " + length() + " which is different from the" 
+                                             + " metric length of " + metricValuesToAdd.length() + " that is being added.");
+    }
+    MetricValues metricValues = _metricValues.computeIfAbsent(metricId, id -> new MetricValues(metricValuesToAdd.length()));
+    metricValues.add(metricValuesToAdd);
+  }
+
+  public void add(AggregatedMetricValues other) {
+    for (Map.Entry<Integer, MetricValues> entry : other.metricValues().entrySet()) {
+      int metricId = entry.getKey();
+      MetricValues otherValuesForMetric = entry.getValue();
+      MetricValues valuesForMetric = _metricValues.computeIfAbsent(metricId, id -> new MetricValues(otherValuesForMetric.length()));
+      if (valuesForMetric.length() != otherValuesForMetric.length()) {
+        throw new IllegalStateException("The two values array has different lengths " + valuesForMetric.length()
+                                            + " and " + otherValuesForMetric.length());
+      }
+      valuesForMetric.add(otherValuesForMetric);
+    }
+  }
+
+  public void subtract(AggregatedMetricValues other) {
+    for (Map.Entry<Integer, MetricValues> entry : other.metricValues().entrySet()) {
+      int metricId = entry.getKey();
+      MetricValues otherValuesForMetric = entry.getValue();
+      MetricValues valuesForMetric = valuesFor(metricId);
+      if (valuesForMetric == null) {
+        throw new IllegalStateException("Cannot subtract a values from a non-existing MetricValues");
+      }
+      if (valuesForMetric.length() != otherValuesForMetric.length()) {
+        throw new IllegalStateException("The two values array has different lengths " + valuesForMetric.length()
+                                            + " and " + otherValuesForMetric.length());
+      }
+      valuesForMetric.subtract(otherValuesForMetric);
+    }
+  }
+  
+  public void clear() {
+    _metricValues.clear();
+  }
+  
+  public void writeTo(OutputStream out) throws IOException {
+    out.write("{%n".getBytes(StandardCharsets.UTF_8));
```

Comment:
OutputStream can be wrapped in OutputStreamWriter which can be constructed with UTF8 output.  This means not having to call getBytes() everytime.